### PR TITLE
gitAndTools.pre-commit: init at 1.10.4

### PR DESCRIPTION
**Pull request by borisbabic**
###### Motivation for this change

Add [pre-commit](https://pre-commit.com/) - a useful tool for managing and sharing git commit hooks.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/aspy.yaml/default.nix [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{ lib, buildPythonPackage, fetchPypi, pythonPackages }:
```
Comment (FRidh) [inline pkgs/development/python-modules/aspy.yaml/default.nix]: no `pythonPackages`, list the individual requirements

File: pkgs/development/python-modules/cfgv/default.nix [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{ lib, buildPythonPackage, fetchPypi, pythonPackages }:
```
Comment (FRidh) [inline pkgs/development/python-modules/cfgv/default.nix]: same

File: pkgs/development/tools/pre-commit/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{ stdenv, pythonPackages, pew }:
+with pythonPackages; buildPythonApplication rec {
+  name = "${pname}-${version}";
```
Comment (FRidh) [inline pkgs/development/tools/pre-commit/default.nix]: The `name` attribute is added by `buildPython*` and should therefore be removed.

File: pkgs/development/tools/pre-commit/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    six
+    toml
+    virtualenv
+  ];
+
+  meta = with stdenv.lib; {
+    description = "A framework for managing and maintaining multi-language pre-commit hooks.";
+    homepage = https://pre-commit.com/;
+    license = licenses.mit;
+    platforms = platforms.all;
```
Comment (FRidh) [inline pkgs/development/tools/pre-commit/default.nix]: drop platforms, is already set

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -186,6 +186,8 @@ in {
 
   asn1crypto = callPackage ../development/python-modules/asn1crypto { };
 
+  aspy.yaml = callPackage ../development/python-modules/aspy.yaml { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: The `.` should be replaced by e.g. `_` (also in the commit message).

Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: preferable even a `-`, so to follow the PEP on normalized names.

Comment (borisbabic) [reply pkgs/top-level/python-packages.nix]: Yeah, I wasn't sure what to do with the period, I changed it now to `-`.

File: pkgs/development/python-modules/aspy.yaml/default.nix [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1ajb97kn044ximkzq2090h4yblrhw77540pwiw345cp7mwzy4xqa";
+  };
+
+  propagatedBuildInputs = [ pyyaml ];
+
+  meta = with lib; {
+    description = "A few extensions to pyyaml.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/aspy.yaml/default.nix]: No period: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes (also for the other descriptions)

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -8420,6 +8420,8 @@ with pkgs;
 
   pyprof2calltree = pythonPackages.callPackage ../development/tools/profiling/pyprof2calltree { };
 
+  pre-commit = pythonPackages.callPackage ../development/tools/pre-commit { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix]: Please don't use `pythonPackages` here, but as argument in the expression.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/pre-commit/default.nix [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    six
+    toml
+    virtualenv
+  ];
+
+  meta = with stdenv.lib; {
+    description = "A framework for managing and maintaining multi-language pre-commit hooks.";
+    homepage = https://pre-commit.com/;
+    license = licenses.mit;
+    maintainers = with maintainers; [ borisbabic ];
```
Comment (dotlambda) [inline pkgs/development/tools/pre-commit/default.nix]: You need to add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix. Please do so in an additional commit.

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: Also, since pre-commit seems to support Python 3, please use that if possible.

Comment (borisbabic) [reply pkgs/top-level/all-packages.nix]: So I don't have to list individual requirements for pre-commit, just for the modules? Or am I misunderstanding something.

I'll use python3. Is the preferred method to do the following here:

```nix
  pre-commit = callPackage ../development/tools/pre-commit {
    pythonPackages = python3Packages;
  };
```

File: pkgs/development/tools/pre-commit/default.nix (hunk 3 above)
Comment (borisbabic) [reply pkgs/development/tools/pre-commit/default.nix]: Yeah, that's redundant. fixed it

File: pkgs/development/tools/pre-commit/default.nix (hunk 4 above)
Comment (borisbabic) [reply pkgs/development/tools/pre-commit/default.nix]: Oh, buildPythonApplication does that?  I dropped it.

File: pkgs/development/python-modules/cfgv/default.nix (hunk 2 above)
Comment (borisbabic) [reply pkgs/development/python-modules/cfgv/default.nix]: I fixed these issues.

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: Ye, correctly. The advantage is that applications can override individual packages.
No, the latter is not preferred. I'd just use `python3` or `python3Packages` as argument.

File: pkgs/development/tools/pre-commit/default.nix [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    cached-property
+    cfgv
+    identify
+    nodeenv
+    six
+    toml
+    virtualenv
+  ];
+
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/tools/pre-commit/default.nix]: Why do you disable tests? Please add a comment.

Comment (borisbabic) [reply pkgs/development/tools/pre-commit/default.nix]: The tests fail on python3 due to a windll dependency. I will add a comment explaining, I should have done that right away :sweat_smile:

**Comment by dotlambda** [comment]
There are two final things to do. Sorry that I didn't think of these earlier:
- The expression for `pre-commit` should be moved to `pkgs/applications/version-management/git-and-tools/pre-commit/default.nix` and called from `pkgs/applications/version-management/git-and-tools/default.nix`.
- All three newly packaged Python modules seem not ship their testsuite in the PyPI tarball. Please add something like this to their expressions:
```
# Tests not included in PyPI tarball
doCheck = false;
```
Anyway, thanks a lot for your work!

**Comment by dotlambda** [comment]
@GrahamcOfBorg build pre-commit

**Comment by dotlambda** [comment]
@GrahamcOfBorg build gitAndTools.pre-commit

**Comment by borisbabic** [comment]
Oops, I forgot to change `pre-commit` to `gitAndTools.pre-commit`.

Thanks everybody for your help and patience!

**Pull request merged**